### PR TITLE
NFSe development

### PR DESCRIPTION
**Pull request by cmsalmeida**
Creating INVOICE_STATE in l10n_br_account, allow us to redefine invoice states, keeping modularity in l10n_br_account_product and l10n_br_account_service. This way we can install only one or both without conflicts.
Added also workflow for NFSe transmission.

### CODE REVIEW

File: l10n_br_account_product/account_invoice.py [hunk 1]
```diff
@@ -30,7 +30,11 @@
     PRODUCT_FISCAL_TYPE_DEFAULT)
 from .product import PRODUCT_ORIGIN
 from .sped.nfe.validator import txt
+from openerp.addons.l10n_br_account.account_invoice import INVOICE_STATE
 
+INVOICE_STATE = INVOICE_STATE + [('sefaz_exception', u'Erro de autorização da Receita'),
+        ('sefaz_cancelled', 'Cancelado no Sefaz'),
+        ('sefaz_denied', 'Denegada no Sefaz')]
```
Comment (danimaribeiro) [inline l10n_br_account_product/account_invoice.py]: 'sefaz_export' Is missing.

File: l10n_br_sale_product/sale.py [hunk 2]
```diff
@@ -367,5 +367,5 @@ def l10n_br_sale_product_prepare_order_line_invoice_line(self, cr, uid, line,
         if line.product_id.fiscal_type == 'product':
             fp_id = line.fiscal_position or result.get('fiscal_position', False) or False
             if fp_id:
-                result['cfop_id'] = fp_id.cfop_id.id
+                result['cfop_id'] = self.pool.get('account.fiscal.position').browse(cr, uid, fp_id).cfop_id.id
```
Comment (danimaribeiro) [inline l10n_br_sale_product/sale.py]: I'm no sure if this work for all the cases.
https://github.com/Trust-Code/l10n-brazil/commit/ab4b781c5fc0ff3c0ffb31c2b4d3b928e8fe25a4

**Comment by rvalyi** [comment]
Ola @cmsalmeida seria legal se puder fazer o PR para a branch da OCA 8.0 https://github.com/OCA/l10n-brazil ou os PR's relacionados porque se vc fica demais na branch odoo-brazil depois vc vai ter um trabalho de migracao importante e a gente vai ter um retrabalho para integrar esses PR's na outra branch. Sendo que vc pode ajudar nos PR's pendente se vc quer aue anda mais rapido. Atts.

File: l10n_br_sale_product/sale.py [hunk 3]
```diff
@@ -365,7 +365,9 @@ def _prepare_order_line_invoice_line(self, cr, uid, line,
     def l10n_br_sale_product_prepare_order_line_invoice_line(self, cr, uid, line,
                                                               result, account_id=False, context=None):
         if line.product_id.fiscal_type == 'product':
-            fp_id = line.fiscal_position or result.get('fiscal_position', False) or False
+            fp_id = line.fiscal_position.id or result.get('fiscal_position', False) or False
```
Comment (danimaribeiro) [inline l10n_br_sale_product/sale.py]: I think 'fiscal_position' inside result generally is an int. But I need to check first.

File: l10n_br_sale/sale.py [hunk 4]
```diff
@@ -194,15 +194,14 @@ def onchange_address_id(self, partner_invoice_id, partner_shipping_id,
...
-            fiscal_category_id=fiscal_category_id)
+            partner_invoice_id, partner_shipping_id, partner_id, company_id, **kwargs)
 
     @api.model
     def _fiscal_position_map(self, result, context=None, **kwargs):
 
         if not context:
             context = {}
-        context.update({'use_domain': ('use_sale', '=', True)})
+        dict(context).update({'use_domain': ('use_sale', '=', True)})
```
Comment (danimaribeiro) [inline l10n_br_sale/sale.py]: Is this correct?

**Comment by danimaribeiro** [comment]
Probably tomorrow I'm going to test this.

File: l10n_br_sale/sale.py (hunk 4 above)
Comment (yk2kus) [inline l10n_br_sale/sale.py]: Yes it is!

**Comment by mileo** [comment]
=D :100: 
@cmsalmeida Você implementou algum código de transmissão também?
[]s

File: l10n_br_account_product/account_invoice.py [hunk 5]
```diff
@@ -836,6 +829,7 @@ def write(self, cr, uid, ids, vals, context=None):
         if not context:
             context = {}
         context.update({'invoice_line_id': ids})
+        vals.update({'invoice_line_tax_id' : [(6, 0 , vals['invoice_line_tax_id'])]})
```
Comment (danimaribeiro) [inline l10n_br_account_product/account_invoice.py]: This line throws me an error:
ValueError: "invoice_line_tax_id" while evaluating
u'action_invoice_create()'

If this fixes an bug you found, better to create a new PR and explain what the problem is.

File: l10n_br_account_product_service/account_invoice_workflow.xml [hunk 6]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (danimaribeiro) [inline l10n_br_account_product_service/account_invoice_workflow.xml]: You shoul add this file to __openerp__.py

Comment (cmsalmeida) [inline l10n_br_account_product_service/account_invoice_workflow.xml]: The workflow is not completed yet, so better to not try to load it now.

Comment (danimaribeiro) [inline l10n_br_account_product_service/account_invoice_workflow.xml]: If you send a PR but still working on it, put the tag [WIP] to it.

Comment (cmsalmeida) [inline l10n_br_account_product_service/account_invoice_workflow.xml]: ok I will, but it's working well without the workflow

**Pull request closed without merging**